[ENH] added piecewise_multinomial function to datagen.py

JonathanBechtel opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/sktime/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
PR #3363 


#### What does this implement/fix? Explain your changes.
In order to complete the unit tests for importing different modules in HMMLearn, additional data generation functions needed to be written for the appropriate modules.  This pull requests added a function called `piecewise_multinomial` that generates multinomial data to be used in unit tests.

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
Making sure the random number generation works in a way that's compatible with the way HMMLearn will ingest it.

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [ X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

miraep8 commented on sktime/annotation/datagen.py:
```diff
@@ -196,6 +196,77 @@ def piecewise_normal(
+    Returns
+    -------
+    data : np.array
+        univariate or multivariate time series as np.array
```
It could be nice (from the user perspective) if you added a detail here about what is the expected size of this array and how that corresponds to the input variables (eg `lengths` and `n_trials`).

miraep8 reviewed (approved):
Just small non-blocking comment.  However - may I also request that if you haven't already added yourself as a contributor to sktime that you do so now? :)  Cheers!

lmmentel commented on sktime/annotation/datagen.py:
```diff
@@ -196,6 +196,77 @@ def piecewise_normal(
+def piecewise_multinomial(
+    n_trials: int,
+    lengths: npt.ArrayLike,
+    pvals: npt.ArrayLike,
```
Here and int he function body you have `pvals` variable name however in the doc string text it's `p_vals`. I guess it makes sense to use `p_vals` since you also chosen `n_trials`. It's a minor thing but consistency often helps.

lmmentel commented on sktime/annotation/datagen.py:
```diff
@@ -196,6 +196,77 @@ def piecewise_normal(
+
+    """
+    rng = check_random_state(random_state)
+    assert len(lengths) == len(pvals)
+    assert all(sum(pval) == 1 for pval in pvals)
+    assert (np.array([len(pval) for pval in pvals]) == len(pvals[0])).all()
```
Would you mind turning `asserts` into `try`/`except` blocks raising human readable errors? The reason is that in some cases of production code there is a way to skip assert statements entirely (see e.g. this [link](https://realpython.com/python-assert-statement/#understanding-the-__debug__-built-in-constant)).  Thanks!

lmmentel commented:
Thanks @JonathanBechtel :raised_hands: Overall looks good. I left a few comments that I hope can make the code a bit more robust.

miraep8 commented:
This should be merge-able once those changes are propagated!  :)  Happy to take another look then!

JonathanBechtel commented:
@lmmentel @miraep8 Made the suggested changes, thank you for the tips.  To recap:
 - updated docstrings to include more detail about output of function
 - removed `assert` statements and changed them to `if` statements w/ a `raise`  statement as suggested
 - changed `pvals` to `p_vals` to make arguments more clear
 
 I see there's a merge conflict in `.all-contributorsrc`.  I added myself as a contributor, which I assume is what's causing it, but I'm not sure if it's best for me to fix that or if it's something that's handled by a code reviewer.
 
 Thanks again for looking at it.

lmmentel reviewed (approved):
LGTM!

miraep8 reviewed (approved):
Looks great to me as well!  Thanks for addressing all the change requests! :) 🎉
